docs: complete the database feature documentation (#5647)
cloudnative-pg/cloudnative-pg: docs/src/declarative_database_management.md

# PostgreSQL Database Management

> CloudNativePG simplifies PostgreSQL database provisioning by automatically

Author: I think this makes more sense if we speak of cluster.
```suggestion
CloudNativePG simplifies PostgreSQL cluster provisioning by automatically
```

Reviewer: I actually meant the database.

Author: So, we say CNPG simplifies database provisioning by creating the `app` database automatically.
An the rest of the document talks about a completely different way of creating databases.

Reviewer 2: The idea is to explain why we always create the app database.

Author: 👍

creating an application database named `app` by default. This default behavior
is explained in the ["Bootstrap an Empty Cluster"](bootstrap.md#bootstrap-an-empty-cluster-initdb)
section.

For more advanced use cases, CloudNativePG introduces **declarative database
management**, which empowers users to define and control the lifecycle of
PostgreSQL databases using the `Database` Custom Resource Definition (CRD).
This method seamlessly integrates with Kubernetes, providing a scalable,
automated, and consistent approach to managing PostgreSQL databases.

---

## Key Concepts

### Scope of Management

!!! Important
    CloudNativePG manages **global objects** in PostgreSQL clusters, such as
    databases, roles, and tablespaces. However, it does **not** manage the content
    of databases (e.g., schemas and tables). For database content, specialized

>     tools or the applications themselves should be used.

Author: Not sure about "application themselves". Which applications? I can make that you're talking about the user payloads, but IMO this is not obvious.
```suggestion
    tools, or the user-defined payloads should be used.
```

    name: cluster-example
```

Note that the Database references a Cluster, on which the database will be
created.
The Database object is managed by the instance manager of the cluster's
primary instance.

In the CRD, the `metadata.name` field represents the name the object
will have in Kubernetes, which is guaranteed to be unique per namespace.
There is also the field `spec.name`, which is the name that will be used for
the database in Postgres.

!!! Note
    Having separate `metadata.name` and `spec.name` makes it possible to have
    two different CloudNativePG clusters in the same namespace, with Databases
    that have the same name in Postgres.

!!! Warning
    While declarative database management in CloudNativePG adheres to the
    PostgreSQL database
    [CREATE](https://www.postgresql.org/docs/current/sql-createdatabase.html)
    and [ALTER](https://www.postgresql.org/docs/current/sql-alterdatabase.html)
    commands, it does not support renaming of databases. Changing the
    `spec.name` in a Database object will be rejected at the Kubernetes level.

### Managing an existing database via a Database manifest

> It is possible to declare a Database object that will handle an existing

Author: ```suggestion
It is possible to declare a Database object that references an existing
```

database. In such case, the Database's fields will be applied using `ALTER`
statements, rather than `CREATE`. There are differences between these two
Postgres commands. In particular, the options accepted by `ALTER` are a subset
of those accepted by `CREATE`.

The database reconciler will transparently handle this on behalf of the user,
making it easy to honor a Database manifest no matter the previous history
of the cluster.

> There is however a difference regarding handling of "immutable" fields: on an

Author: ```suggestion
There is however a difference regarding the handling of "immutable" fields: on an
```

> There is however a difference regarding handling of "immutable" fields: on an

Author: ```suggestion
There is however a difference regarding the handling of "immutable" fields: on an
```

existing Database object, any modification of the immutable fields will
be rejected at the Kubernetes level.
On a newly declared Database manifest that references an existing database, the
immutable fields will simply be ignored, as they are not valid options in the
`ALTER DATABASE` command.

!!! Warning
    If a Database manifest references an existing database, any fields in the

>     manifest that cannot be set in `ALTER DATABASE` will be dropped silently.

Author: ```suggestion
    manifest that cannot be set in `ALTER DATABASE` will be ignored.
```

    name: cluster-example
```

Note that the Database references a Cluster, on which the database will be
created.
The Database object is managed by the instance manager of the cluster's

> primary instance. Declarative database management is not supported in replica

Author: Ah, interesting. OK on one hand, the Database object reconciliation would fail, and would be re-queued, so if eventually we promoted the cluster, the database reconciliation would proceed.
Now, in a different direction: this feature is not so much doing "management" as doing declarative creation. Once a manifest is successfully applied, it's over and done until we alter it (and the `generation` is incremented).
So, to understand your use case better: if we create Database Foo in the primary, then it will be replicated in the replica cluster, no? So having a copy of the Database manifest in the replica cluster would not be of any use.

> primary instance. Declarative database management is not supported in replica

Author: I've rewritten this, I hope it's now clearer.

the user to take appropriate steps.

For example, applying the following Database manifest:

```yaml
apiVersion: postgresql.cnpg.io/v1
  kind: Database
  metadata:
    name: db-icu
spec:
  name: declarative-icu
  owner: app
  encoding: UTF8
  locale_provider: icu
  icu_locale: en
  icu_rules: fr
  template: template0
  cluster:
    name: cluster-example
```

on a cluster running Postgres 14 will result in an error message in the
Database object's status:

```yaml
[...]
status:
  ready: false
  error: option "locale_provider" not recognized
```

The rationale is that this is exactly what will happen if you attempt to create
a database directly on the `psql` command line. The database reconciler aims
at transparency.

## Making direct changes in Postgres

It is possible to make changes to a database that was created or managed with a
Database object, directly on Postgres, for example by issuing commands on
`psql`.

The fields `observedGeneration` and `generation` described above will ensure
that once a Database has been reconciled to its defined `generation`, it will
not be re-applied by the instance manager. Therefore, your manual changes will
not be rolled back inadvertently.

> !!! Note
>     A Database manifest is applied to the fullest. A field included in a
>     manifest may override a value that had been written previously on
>     the database in Postgres
> 

Author: ```suggestion
```
